src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,253 @@
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	waitDuration	= time.Second * 3
+	waitFactor	= 2
+	waitJitter	= 0.5
+	waitSteps	= 5
+	waitCap		= time.Minute * 5
+)
+
+var backoff = wait.Backoff{
+	Duration:	waitDuration,
+	Factor:		waitFactor,
+	Jitter:		waitJitter,
+	Steps:		waitSteps,
+	Cap:		waitCap,
+}
+
+func (r *StatefulSetResource) updateStsImage(
+	ctx context.Context, sts *appsv1.StatefulSet,
+) error {
+	rpContainer, err := findContainer(sts.Spec.Template.Spec.Containers, redpandaContainerName)
+	if err != nil {
+		return err
+	}
+
+	newImage := fmt.Sprintf("%s:%s", r.pandaCluster.Spec.Image, r.pandaCluster.Spec.Version)
+	if rpContainer.Image != newImage {
+		r.logger.Info("Image has changed", "cluster image", newImage, "container image", rpContainer.Image)
+
+		stsSpec := sts.Spec.Template.Spec
+
+		if err := modifyContainerImage(stsSpec.InitContainers, configuratorContainerName, newImage); err != nil {
+			return err
+		}
+
+		if err := modifyContainerImage(stsSpec.Containers, redpandaContainerName, newImage); err != nil {
+			return err
+		}
+
+		if err := r.partitionUpdateImage(ctx, sts, newImage); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string,
+) error {
+	r.logger.Info("Start statefulset image update")
+
+	// When a StatefulSet's partition number is set to `i`, only Pods with ordinal
+	// greater than or equal to `i` will be updated.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#partitions
+	for ordinal := *sts.Spec.Replicas - 1; ordinal >= 0; ordinal-- {
+		// Update() on statefulset has not been called yet in this run, however,
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
+			continue
+		}
+
+		// Continue only if error is due to pod not ready, or unchanged image.
+		if !errors.Is(poderr, errContainerHasWrongImage) && !errors.Is(poderr, errPodNotReady) {
+			return poderr
+		}
+
+		// Ideally, at this point we'd wait until all Pods are in ready state.
+		// However, if something went wrong with a previous image change and the pod
+		// is in bad state, e.g., due to wrong or erroneous image, the pod will
+		// keep restarting and require manual intervention. Instead, for now we
+		// skip the readiness requirement and "force" the image update.
+
+		if err := r.rollingUpdatePartition(ctx, sts, ordinal); err != nil {
+			return err
+		}
+
+		r.logger.Info("Waiting for pod to complete updating", "partition:", ordinal)
+
+		if err := r.waitUntilPodImageIsVerified(ctx, sts, newImage, ordinal); err != nil {
```

Comment:
one concern I have about this approach that the whole update process blocks the reconciliation worker which is problematic and usually not ideal approach because potentially other changes can stall depending on size of the cluster and amount of workers you have.

Have you thought about making this asynchronous? Could we perhaps watch on pod updates and proceed based on that?

Reply:
The wait was removed. Resolving this.